File: server_data/contests/Contest3/replay/validator.py
#!/usr/bin/env python3
import sys
from collections import Counter

# This validator receives file paths as command-line arguments from the judge engine.
# sys.argv[1]: Path to the problem's input file.
# sys.argv[2]: Path to the user's generated output file.
# sys.argv[3]: Path to the problem's official/expected output file. (Not used in this validator)

def judge_error(message: str):
    """
    Exits with code 2 for judge errors (e.g., malformed input, validator crash).
    This signals an "Internal Error" verdict.
    """
    print(f"Judge Error: {message}", file=sys.stderr)
    sys.exit(2)


def wrong_answer(message: str):
    """
    Exits with code 1 for a "Wrong Answer" verdict.
    """
    # print(f"WA: {message}", file=sys.stderr)
    sys.exit(1)


def accept():
    """
    Exits with code 0 for an "Accepted" verdict.
    """
    sys.exit(0)


def main():
    if len(sys.argv) < 4:
        judge_error(f"Validator was called with insufficient arguments. Got {len(sys.argv) - 1}, expected 3.")

    problem_input_path = sys.argv[1]
    user_output_path = sys.argv[2]
    # expected_output_path = sys.argv[3]

    try:
        with open(problem_input_path, 'r') as f:
            # The input format is a line of space-separated integers
            expected_freq_list = list(map(int, f.read().strip().split()))
        if len(expected_freq_list) != 26:
            judge_error(f"Problem input file has {len(expected_freq_list)} numbers, expected 26.")
    except (IOError, ValueError) as e:
        judge_error(f"Could not read or parse the problem's input file: {e}")

    # --- 3. Read User's Output ---
    user_output_str = ""

    try:
        with open(user_output_path, 'r') as f:
            user_output_str = f.read().strip()
    except IOError:
        # If the user's program produces no output file, treat it as an empty string.
        pass

    odd_freq_count = 0
    for freq in expected_freq_list:
        if freq < 0:
            judge_error("Problem input contains negative frequency.")
        if freq % 2 != 0:
            odd_freq_count += 1

    # Validation
    is_possible = (odd_freq_count <= 1)

    if not is_possible:
        if user_output_str == "-1":
            accept()
        else:
            wrong_answer(f"A palindrome is impossible, but user printed '{user_output_str}' instead of '-1'.")
    else:
        if user_output_str == "-1":
            wrong_answer("A palindrome is possible, but user printed '-1'.")

        if user_output_str != user_output_str[::-1]:
            wrong_answer("The output string is not a palindrome.")

        user_freq_map = Counter(user_output_str)

        user_freq_list = [0] * 26
        for char, count in user_freq_map.items():
            if 'a' <= char <= 'z':
                user_freq_list[ord(char) - ord('a')] = count
            else:
                wrong_answer(f"Output contains an invalid character: '{char}'.")

        if user_freq_list != expected_freq_list:
            wrong_answer("The letters or their frequencies in the output do not match the problem input.")

        accept()


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        # Catch any unexpected crashes in the validator itself.
        judge_error(f"An uncaught exception occurred in the validator: {e}")

File: server_data/contests/Contest3/chronos/validator.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <cctype>

// This validator receives file paths as command-line arguments from the judge engine.
// argv[1]: Path to the problem's input file. (Not used in this validator)
// argv[2]: Path to the user's generated output file.
// argv[3]: Path to the problem's official/expected output file.

// Exit with code 2 for judge errors (e.g., file read errors).
// This signals an "Internal Error" verdict.
void judge_error(const std::string& msg) {
    std::cerr << "Judge Error: " << msg << std::endl;
    exit(2);
}

// Exit with code 1 for a "Wrong Answer" verdict.
void wrong_answer() {
    exit(1);
}

// Exit with code 0 for an "Accepted" verdict.
void accept() {
    exit(0);
}

// Helper function to convert a string to lowercase.
std::string to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c){ return std::tolower(c); });
    return s;
}

std::string trim(const std::string& s) {
    const std::string whitespace = " \t\n\r\f\v";
    size_t first = s.find_first_not_of(whitespace);
    if (std::string::npos == first) {
        return "";
    }
    size_t last = s.find_last_not_of(whitespace);
    return s.substr(first, (last - first + 1));
}


int main(int argc, char* argv[]) {
    if (argc < 4) {
        judge_error("Validator was called with insufficient arguments.");
    }

//    std::string problem_input_path = argv[1];
    std::string user_output_path = argv[2];
    std::string expected_output_path = argv[3];

    std::ifstream user_output_file(user_output_path);
    std::string user_output_str = "";
    if (user_output_file.is_open()) {
        user_output_str.assign(
            (std::istreambuf_iterator<char>(user_output_file)),
            (std::istreambuf_iterator<char>())
        );
        user_output_file.close();
    } else {
        // If the user's program produces no output file, treat it as an empty string.
    }

    std::ifstream expected_output_file(expected_output_path);
    std::string expected_output_str;
    if (expected_output_file.is_open()) {
        expected_output_str.assign(
            (std::istreambuf_iterator<char>(expected_output_file)),
            (std::istreambuf_iterator<char>())
        );
        expected_output_file.close();
    } else {
        judge_error("Could not open the official expected output file.");
    }

    // Validation
    std::string user_trimmed = trim(user_output_str);
    std::string expected_trimmed = trim(expected_output_str);

    if (to_lower(user_trimmed) == to_lower(expected_trimmed)) {
        accept();
    } else {
        wrong_answer();
    }

    return 0;
}